pkg/flow/module.go

```diff
@@ -2,19 +2,51 @@ package flow
 
 import (
 	"context"
+	"fmt"
 	"net/http"
 	"path"
 	"sync"
 
 	"github.com/gorilla/mux"
 	"github.com/grafana/agent/component"
 	"github.com/grafana/agent/pkg/cluster"
+	"github.com/grafana/agent/pkg/flow/internal/controller"
 	"github.com/grafana/agent/pkg/flow/logging"
+	"github.com/grafana/agent/pkg/flow/tracing"
 	"github.com/grafana/agent/web/api"
 	"github.com/prometheus/client_golang/prometheus"
-	"go.opentelemetry.io/otel/trace"
 )
 
+type moduleController struct {
+	o   *moduleControllerOptions
+	ids map[string]struct{}
+}
+
+var (
+	_ component.ModuleController = (*moduleController)(nil)
+)
+
+// newModuleController is the entrypoint into creating module instances.
+func newModuleController(o *moduleControllerOptions) component.ModuleController {
+	return &moduleController{
+		o:   o,
+		ids: map[string]struct{}{},
+	}
+}
+
+// NewModule creates a new, unstarted Module.
+func (m *moduleController) NewModule(id string, export component.ExportFunc) (component.Module, error) {
+	if _, found := m.ids[id]; found {
+		return nil, fmt.Errorf("id %s already exists", id)
+	}
+	m.ids[id] = struct{}{}
```

Comment:
Unless I'm missing something, this leaks now. If you remove a module loader component from a config and then bring it back, it will fail once the config is reloaded.